website/blog/2021-08-05-win32component.md

```diff
@@ -0,0 +1,416 @@
+- We subscribe to two events: `RequestReceived`, which is triggered when the React Native application sends a message; `ServiceClosed`, which is triggered when the App Service channel is shut down.
+
+Then we use the `OpenAsync()` method to open the connection. If it isn't successful, there's no need for this process to stay alive, so we call `Set()` on the `AutoResetEvent` object, so that the `Main()` method can terminate.
+
+Let's see now the implementation of the two event handlers. The first one, `ServiceClosed`, is the easiest one:
+
+```csharp
+private static void Connection_ServiceClosed(AppServiceConnection sender, AppServiceClosedEventArgs args)
+{
+    //when the connection with the App Service is closed, we terminate the Win32 process
+    appServiceExit.Set();
+}
+```
+
+If the communication with the App Service drops, we simply call `Set()` on the `AutoResetEvent` object so that the process can terminate.
+
+The second one, `RequestReceived`, instead, is the one where the communication with the React Native application happens:
+
+```csharp
+private static async void Connection_RequestReceived(AppServiceConnection sender, AppServiceRequestReceivedEventArgs args)
+{
+    string key = args.Request.Message["RegistryKeyName"].ToString();
+
+    var hive = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Windows NT\CurrentVersion", false);
+    string value = hive.GetValue(key).ToString();
+
+    ValueSet valueSet = new ValueSet
+    {
+        { "RegistryKeyValue", value }
+    };
+
+    await args.Request.SendResponseAsync(valueSet);
+}
+```
+
+From the `Request.Message` dictionary of the event arguments, we can retrieve any information passed by the main React Native application. In our scenario, the name of the requested registry key will be stored in an item with `RegistryKeyName` as key.
+We use this information to connect to the registry (using the `Registry` APIs) and to retrieve this key from `SOFTWARE\Microsoft\Windows NT\CurrentVersion`, which is the hive that contains all the information about the currently installed version of Windows.
+
+Once we have retrieved the key, we stored it a new message (the `ValueSet` object) with key `RegistryKeyValue`. In the end, we send it back using the `SendResponseAsync()` method, so that the caller (in our case, the native module) can use it.
+
+### The React Native for Windows application
+
+When the classic Windows application invokes the `OpenAsync()` method to connect to the App Service, the host UWP app will wake up to respond (remember that we're using an in-proc App Service). This scenario is handled by a special event called `OnBackgroundActivated`, which you can subscribe in the `App` class of the host app. This is the implementation:
```

Comment:
i'm confused by the mention of in-proc app service when the win32 app is running in a separate process

Reply:
If you think it's confusing the mention here is not really needed. The full story is that the App Service is running in-proc in our scenario, which means that the app itself is responsible for managing the background activation. App Services, however, can be also out-proc, which means that the background activation is handled by a dedicated Windows Runtime Component, which is executed by another process (backgroundTaskHost.exe).